packages/web-extension/src/messaging/remoteApi.ts

```diff
@@ -242,22 +257,28 @@ export const bindObservableChannels = <API extends object>(
 };
 
 /**
- * Bind an API object to handle messages from other parts of the extension.
- * This can only used once per channelName per process.
+ * Bind an API object emitted by `api$` observable to handle messages from other parts of the extension.
+ * - This can only used once per channelName per process.
+ * - Changing source `api` object is possible by emitting it from the `api$` observable.
+ * - Subscriptions to observable properties are kept active while `api$` is still active. This allows changing the
+ *   observed `api` object without having to resubscribe the properties.
+ * - Observable properties are completed only when `api$` completes.
+ * - Methods returning `Promises` will await until the first `api` object is emitted.
  *
- * NOTE: All Observables are subscribed when this function is called.
+ * NOTE: All Observables are subscribed when this function is called and an `api` object is emitted by `api$`.
  * Caches and replays (1) last emission upon remote subscription (unless item === null).
  *
  * In addition to errors thrown by the underlying API, methods can throw TypeError
  *
  * @returns object that can be used to shutdown all ports (shuts down 'messenger' dependency)
  */
 export const exposeMessengerApi = <API extends object>(
-  { api, properties }: ExposeApiProps<API>,
+  { api$, properties }: ExposeApiProps<API>,
   dependencies: MessengerApiDependencies
 ): Shutdown => {
-  const observableChannelsSubscription = bindObservableChannels({ api, properties }, dependencies);
-  const nestedObjChannelsSubscription = bindNestedObjChannels({ api, properties }, dependencies);
+  const observableChannelsSubscription = bindObservableChannels({ api$, properties }, dependencies);
+  const nestedObjChannelsSubscription = bindNestedObjChannels({ api$, properties }, dependencies);
+  const apiTracker$ = new TrackerSubject(api$);
```

Comment:
Now it can be completed in 2 ways: completing `api$`, or calling `shutdown()`. I think it would be a little better if there was only 1 way of doing that. This would rely only on `shutdown()`:
```suggestion
  const apiTracker$ = new TrackerSubject(concat(api$, NEVER));
  const observableChannelsSubscription = bindObservableChannels({ api$: apiTracker$, properties }, dependencies);
  const nestedObjChannelsSubscription = bindNestedObjChannels({ api$: apiTracker$, properties }, dependencies);
```

Also there might be edge cases where object emitted by api$ is shutdown before this. Maybe we should extend this so that `api$: Observable<API | null>` to be able to remove the local source (so that any calls and subscriptions would await for a new api obj before resolving, instead of using a destroyed obj). This is actually a legit use case with wallets: deleting a wallet and creating a new one.

Reply:
Added shutdown lifecycle in [b0b543f](https://github.com/input-output-hk/cardano-js-sdk/pull/489/commits/b0b543f4e0bc5b38be16c67535de6f0a34b5a504).
Handling `null` objects is not done yet. I am working to add better unit tests coverage for the remoteApi in general.

Reply:
Unit tests added in [0e00207](https://github.com/input-output-hk/cardano-js-sdk/pull/489/commits/0e002077dc7bb334e6b530436a071ae66e1e9133).

Reply:
- null api handling to disable monitoring is implemented in: [9972fa4](https://github.com/input-output-hk/cardano-js-sdk/pull/489/commits/9972fa41de215c74487d1a50f71b1e5639d7336c).
- split unit tests into 2 files (unit & integration) ([aaa7d83](https://github.com/input-output-hk/cardano-js-sdk/pull/489/commits/aaa7d83eaa15bb88e94578e82d46e80db9214064), [7cadb7b](https://github.com/input-output-hk/cardano-js-sdk/pull/489/commits/7cadb7ba61c3e265940b7c1754fa7bc068282905))
- unit tests have been refactored to run same test suite on simple/flat object and nested object ([b14c58b](https://github.com/input-output-hk/cardano-js-sdk/pull/489/commits/b14c58bab0622dfe3d44d948bbdd19120e10f649))